### ISSUE TITLE
Escaped braces aren't handled properly

### ISSUE THREAD

**Issue by micimize** [bmatcuk/doublestar#56]
It seems that `\{` and `\}` are not handled properly in both pattern patching and path evaluation:
* The pattern `**/log file \{????-??-??\}.txt` will fail with a syntax error
* The input path `src/logs/log file {2021-05-11}.txt` will fail with any pattern (this is a valid path)

If I remove both from my test the other patterns and path combinations work fine. This does not seem to be the case for brackets, e.g. `src/logs/log file [2021-05-11].txt`
and `**/log file [????-??-??].txt`. I am using `doublestar.Match`.

**Comment by bmatcuk** [comment]
Hmm, are you using doublestar v4? What OS are you on? How are you building your pattern? You'll need to double the backslashes to get them to stay in the string, ie `"**/log file \\{????-??-??\\}.txt"`.

I cannot reproduce this issue using v4 on macOS.

**Comment by micimize** [comment]
I think I was maybe mistaken as to the exact issue – there is an issue with both `\[` and `\{` escaping. 
If there is a match, they succeed, but if not, they return a syntax error. See https://play.golang.org/p/Saso864XooB

<details>
<summary>copied for reference</summary>

```go
package main

import (
	"fmt"

	"github.com/bmatcuk/doublestar/v4"
)

func main() {
	patterns := []string{
		//`\{????-??-??\}.txt`,
		//`\[????-??-??\].txt`,
		`\{*\}.txt`,
		`\[*\].txt`,
	}
	names := []string{
		`{2021-05-01}.txt`,
		`[2021-05-01].txt`,
		`normal text`,
	}
	for _, p := range patterns {
		for _, n := range names {
			match, err := doublestar.Match(p, n)
			fmt.Println(p, n, match, err)
		}
	}

}

/* output
\{*\}.txt {2021-05-01}.txt true <nil>
\{*\}.txt [2021-05-01].txt false syntax error in pattern
\{*\}.txt normal text false syntax error in pattern
\[*\].txt {2021-05-01}.txt false syntax error in pattern
\[*\].txt [2021-05-01].txt true <nil>
\[*\].txt normal text false syntax error in pattern
*/

</details>

**Comment by micimize** [comment]
@bmatcuk I found the issue – every iteration of the loop consuming pattern validates the remaining pattern:
https://github.com/bmatcuk/doublestar/blob/1fd2e2d0a06751f3a9a9174c29b6bc1f215f22b1/match.go#L279-L281

The `default` fallthrough after `'\\'` is matched breaks before bumping `patIdx` when a match is rejected, resulting in an invalid pattern at validation time:
https://github.com/bmatcuk/doublestar/blob/1fd2e2d0a06751f3a9a9174c29b6bc1f215f22b1/match.go#L241-L245

Why validate the pattern every loop? `doValidatePattern` is called `nameLen * patLen` times this way

**Comment by bmatcuk** [comment]
Hey @micimize! I'm so sorry I didn't see your reply back in May! I never got a notification from github that you replied! Your test cases were most helpful and I was able to fix the bug fairly quickly. I only wish I would have seen your reply a month ago so I could have fixed it then :cry:

Anyway, I just released v4.0.2 with the fix. Let me know if you have any trouble with it.

By the way, `doValidatePattern` is only called if all else fails. Everything in the loop above it will either `return` or `continue` the loop before reaching that point. So, it will only reach that if the pattern has been valid up until now, but a match could not be made, and neither `**` or `*` backtracking is possible. At that point, we already know we're returning false, we just want to know if we're also returning an error.
